fpga_interchange/site_router.cc

```diff
@@ -1070,6 +1070,52 @@ static void block_lut_outputs(SiteArch *site_arch,
     }
 }
 
+// Recursively visit downhill PIPs until a SITE_PORT_SINK is reached.
+// Marks all PIPs for all valid paths.
+static bool visit_downhill_pips(const SiteArch *site_arch, const SiteWire &site_wire, std::vector<PipId> &valid_pips) {
+    bool valid_path_exists = false;
+    for (SitePip site_pip : site_arch->getPipsDownhill(site_wire)) {
+        const SiteWire &dst_wire = site_arch->getPipDstWire(site_pip);
+        if (dst_wire.type == SiteWire::SITE_PORT_SINK) {
+            valid_pips.push_back(site_pip.pip);
+            return true;
+        }
+
+        bool path_ok = visit_downhill_pips(site_arch, dst_wire, valid_pips);
+        valid_path_exists |= path_ok;
+
+        if (path_ok) {
+            valid_pips.push_back(site_pip.pip);
+        }
+    }
+
+    return valid_path_exists;
+}
+
+// Checks all downhill PIPs starting from driver wires.
+// All valid PIPs are stored and returned in a vector.
+static std::vector<PipId> check_downhill_pips(Context *ctx, const SiteArch *site_arch) {
+    auto &cells_in_site = site_arch->site_info->cells_in_site;
+
+    std::vector<PipId> valid_pips;
+    for (auto &net_pair : site_arch->nets) {
+        NetInfo *net = net_pair.first;
+        const SiteNetInfo *site_net = &net_pair.second;
+
+        if (net->driver.cell && cells_in_site.count(net->driver.cell)) {
+            const SiteWire &site_wire = site_net->driver;
+
+            bool result = visit_downhill_pips(site_arch, site_wire, valid_pips);
+
+            // There needs to be at least one path that leads
+            // from a cell driver to a site output sink
+            NPNR_ASSERT(result);
```

Comment:
this isn't necessarily true in two cases:
 - 1. all of the users of the net are contained within the site
 - 2. the net has no users (this is a trivial case and just needs `&& !net->users.empty()` added to the if statement)

Reply:
Added fix for case ii.

For i. I might be missing something here, but I guess that, if a site wire corresponding to a net driver got assigned to a net, that would mean that that net was already handled during site routing, therefore there is no need to check for it's valid PIPs.

Reply:
> For i. I might be missing something here, but I guess that, if a site wire corresponding to a net driver got assigned to a net, that would mean that that net was already handled during site routing, therefore there is no need to check for it's valid PIPs

It's true that we don't care about the set of valid PIPs; but that means we also need to ignore the net and not have an assertion that is likely to fail (because it doesn't matter if we find PIPs to a site pin or not).

Reply:
Hmm, right, I wonder if this assertion needs to be there at all at this point. As the router would fail as well if there are no PIPs to route to a site pin, with or without this assertion. Given that there are drivers within a site that may not have routes to a site pin at all, and these drivers are not handled by the router (I think).

Reply:
Yeah I think it's safest just to remove the assertion for now.